ros/ros2/OdometryServer.cpp

```diff
@@ -80,111 +80,94 @@ OdometryServer::OdometryServer(const rclcpp::NodeOptions &options)
-            *this, tf2_ros::StaticBroadcasterQoS(), options);
-
-        geometry_msgs::msg::TransformStamped alias_transform_msg;
-        alias_transform_msg.header.stamp = this->get_clock()->now();
-        alias_transform_msg.transform.translation.x = 0.0;
-        alias_transform_msg.transform.translation.y = 0.0;
-        alias_transform_msg.transform.translation.z = 0.0;
-        alias_transform_msg.transform.rotation.x = 0.0;
-        alias_transform_msg.transform.rotation.y = 0.0;
-        alias_transform_msg.transform.rotation.z = 0.0;
-        alias_transform_msg.transform.rotation.w = 1.0;
-        alias_transform_msg.header.frame_id = child_frame_;
-        alias_transform_msg.child_frame_id = "base_link";
-        br->sendTransform(alias_transform_msg);
-    }
+    tf2_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
+    tf2_buffer_->setUsingDedicatedThread(true);
+    tf2_listener_ = std::make_unique<tf2_ros::TransformListener>(*tf2_buffer_);
 
     RCLCPP_INFO(this->get_logger(), "KISS-ICP ROS 2 odometry node initialized");
 }
 
+Sophus::SE3d OdometryServer::CloudToBaseTf(const std::string &cloud_frame_id) const {
+    std::string err_msg;
+    if (tf2_buffer_->_frameExists(base_frame_) &&     //
+        tf2_buffer_->_frameExists(cloud_frame_id) &&  //
+        tf2_buffer_->canTransform(cloud_frame_id, base_frame_, tf2::TimePointZero, &err_msg)) {
+        try {
+            auto tf = tf2_buffer_->lookupTransform(cloud_frame_id, base_frame_, tf2::TimePointZero);
+            return tf2::transformToSophus(tf);
+        } catch (tf2::TransformException &ex) {
+            RCLCPP_WARN(this->get_logger(), "%s", ex.what());
+        }
+    }
+    RCLCPP_ERROR(this->get_logger(), "Failed to find tf. Reason=%s", err_msg.c_str());
+    return {};
+}
+
 void OdometryServer::RegisterFrame(const sensor_msgs::msg::PointCloud2::ConstSharedPtr &msg) {
+    const auto cloud_frame_id = msg->header.frame_id;
     const auto points = PointCloud2ToEigen(msg);
     const auto timestamps = [&]() -> std::vector<double> {
         if (!config_.deskew) return {};
         return GetTimestamps(msg);
     }();
+    const auto egocentric_estimation = (base_frame_.empty() || base_frame_ == cloud_frame_id);
 
     // Register frame, main entry point to KISS-ICP pipeline
     const auto &[frame, keypoints] = odometry_.RegisterFrame(points, timestamps);
 
-    // PublishPose
-    const auto pose = odometry_.poses().back();
+    // Compute the pose using KISS, ego-centric to the LiDAR
+    const Sophus::SE3d kiss_pose = odometry_.poses().back();
+
+    // If necessary, transform the ego-centric pose to the specified base_link/base_footprint frame
+    const auto pose = [&]() -> Sophus::SE3d {
+        if (egocentric_estimation) return kiss_pose;
+        const Sophus::SE3d cloud2base = CloudToBaseTf(cloud_frame_id);
+        return cloud2base.inverse() * kiss_pose * cloud2base;
```

Comment:
This we have to modify to correctly handle a non-static `cloud2base` transformation.

Since `kiss_pose` is expressed w.r.t the first frame of the sensor at time 0, `cloud2base` transforms between sensor and base link in at the current timestep (which is in case of a non-static tf not necessarily the same as the one at time 0).

Should be something like

```cpp
return cloud2base_ref.inverse() * kiss_pose * cloud2base_current;
```

Reply:
I agree, 

We could convert the wrapper to a lifecycle node, and in this case, while the node is being created, fetch the transform there and take it as the "zero" time. Since it does not always start from zero. For example, when I do some experiments, the stack is already running, and therefore the time 0 was long ago.

I will open this in a separate thread for fixing it later ;)